packages/core/src/build/service.ts

```diff
@@ -278,6 +320,80 @@ export const start = async (
   return buildResult;
 };
 
+export const startServer = async (
+  buildService: Service,
+  {
+    watch,
+    onBuild,
+  }:
+    | { watch: true; onBuild: (buildResult: BuildResultServer) => void }
+    | { watch: false; onBuild?: never },
+): Promise<BuildResultServer> => {
+  const { common } = buildService;
+
+  const serverResult = await executeServer(buildService);
+
+  if (serverResult.status === "error") {
+    return { status: "error", error: serverResult.error };
+  }
+
+  const buildResult = validateAndBuildServer(buildService, serverResult);
+
+  // If watch is false (`ponder start` or `ponder serve`),
+  // don't register  any event handlers on the watcher.
+  if (watch) {
+    // Define the directories and files to ignore
+    const ignoredDirs = [common.options.generatedDir, common.options.ponderDir];
+    const ignoredFiles = [
+      path.join(common.options.rootDir, "ponder-env.d.ts"),
+      path.join(common.options.rootDir, ".env.local"),
+    ];
+
+    const isFileIgnored = (filePath: string) => {
+      const isInIgnoredDir = ignoredDirs.some((dir) => {
+        const rel = path.relative(dir, filePath);
+        return !rel.startsWith("..") && !path.isAbsolute(rel);
+      });
+
+      const isIgnoredFile = ignoredFiles.includes(filePath);
+      return isInIgnoredDir || isIgnoredFile;
+    };
+
+    const onFileChange = async (file: string) => {
+      if (isFileIgnored(file)) return;
+
+      const files = glob.sync(buildService.serverPattern);
+
+      if (files.includes(file) === false) return;
+
+      // Invalidate all server modules.
+      buildService.viteNodeRunner.moduleCache.invalidateDepTree(files);
+
+      common.logger.info({
+        service: "build",
+        msg: `Hot reload ${files
+          .map((f) => `'${path.relative(common.options.rootDir, f)}'`)
+          .join(", ")}`,
+      });
+
+      buildService.viteNodeRunner.moduleCache.deleteByModuleId("@/generated");
```

Comment:
Could these invalidations happen within `executeServer`? I think I'd find that easier to follow.